### DISCUSSION TITLE
How to Roll Back to a Version - Easy Explanation For Non Technical Folks

### DISCUSSION THREAD

**Discussion by pw405** [lshqqytiger/stable-diffusion-webui-directml#376] [General]
You can easily revert to the previous version that was working fine if you can't launch the new version to launch like myself and many others (everybody?) seem to be having trouble with. 

I've never used GitHub before, so this is all new to me as well.  GitHub's Discussion board UI is a complete mess, or I could be an idiot.  Unsure.  Either way, trying to chase breadcrumbs of responses across different topics is a PITA!!  

Until the new version can be launched, you can choose to roll back a version if you like.   Very easy.  I'll explain as if you've never touched a line of code in your life!

Tip #1: Comment syntax.  "Commenting out" just means "ignore this line of code". The Webui-user.**bat** is a batch script.  You can comment out a line by using two colons  ::  or the letters "rem".  Example in Notepad++ using batch script coloring scheme just to show you the two commenting out methods compared to the same lines that are not commented out.  

![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/20309609/846cac1f-cdb3-4ee9-8f8b-8399f03029df)
Just make a mental note for later. 

Rollback steps.   
(Note - make a backup copy of webui-user.bat if you're worried about breaking it.  Just click it in File Explorer, control + C, then Control + V to copy/paste a new version before you make edits .) 

1.  Find the hash number of the version ("commits", in GitHub parlance)  you want to use.  Click on Commits from the main page:
![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/20309609/7b23bfbf-c6a1-411e-aa06-126c0650eb42)

For me, the version on January 13th was working fine.  You can copy that hash to your clipboard by clicking the two boxes:

![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/20309609/183da1ca-2802-483e-a532-30ff4e839e1e)

2.  Use "git checkout".  Comment out any "git pull" you might have in your webui-user.bat !  Keep the command line arguments you were using for that version.  Mine looks like this:
`git checkout d500e58a65d99bfaa9c7bb0da6c3eb5704fadf25`
![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/20309609/43714dcb-86be-42a4-9018-fcd87073a825)

3.  Save the file, and then run it like you normally would.   This should allow you to launch the WebUI.  After the first time you launch the batch file, you can comment out the line with `git checkout d500e58a65d99bfaa9c7bb0da6c3eb5704fadf25`  put `rem `or `::` in front of the line.  This can be done while WebUI is running or later.  Save it.

![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/20309609/d14691b9-8024-471d-8435-b281dc18d071)

You should see the console confirm  a "Detached State" message:
![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/20309609/3691c6c0-a198-42e2-89b1-ebef00140866)

Keep in mind, any new updates will **not automatically download** when your batch file is in this state.   Great, how do I get back to  the way it was?!

4. Return to the old ways.  Just as easy as last time, comment out or delete the git checkout line.  Instead, use `git checkout master`  Save the changes to the file. Here's what mine looks like:
![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/20309609/6af7f21f-f6ef-4430-bb06-66dd821bdd66)

Now, launch the webui-user.bat file.  You should see a message saying you switched to the master branch:
![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/20309609/e8e9baf8-1378-4467-9213-dc0603557480)

Close the console out after it launches.  Now back in the batch file, let's delete or comment out `git checkout master` and just use good ol' `git pull`  this time.  Launch it and you should be back back to normal.   Here's what mine looks like:
![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/20309609/f9908865-f89d-4459-bc58-5ab0aecf1817)

NOTE - On the last steps when returning to normal, I THINK you can just do: 
```
git checkout master
git pull
```
all at once, but not sure if that will cause problems.   Like I said, this is the first GitHub project I've ever used and don't have much experience with it. PLEASE correct any part of this in the comments so that other users will have accurate info.  Or if you know of a better/easier way, let us know!  

Hope this helps those who are having trouble with new version so we can enjoy and leave the author to bug fixes!

**Comment by Reizer88** [comment]
Yes the update seems having some error for us yesterday, after updating my SD i get error, webui can open in browser but getting error n not work. After that i try to delete venv folder etc n run webui-user i still get a error, so after update for api fix has launched i decide to do a new clean installation on a new directory, doing git clone, edit webui-user to add arguments, moving all my checkpoints etc to new installation directory, run webui-user and finally it work again.
This is my arguments
`@echo off

set PYTHON=
set GIT=
set VENV_DIR=
set COMMANDLINE_ARGS=--use-directml --medvram --opt-split-attention --opt-split-attention-v1 --precision full --no-half --no-half-vae --opt-sub-quad-attention --sub-quad-q-chunk-size 512 --sub-quad-kv-chunk-size 512 --sub-quad-chunk-threshold 80 --precision full --always-batch-cond-uncond --disable-nan-check --upcast-sampling --theme dark --autolaunch

call webui.bat
`   
I use RX6600 8 GB, 32GB RAM, GPU Driver Version 23.12.1

![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/98239147/bb124bfe-06bb-478b-bb50-91422e52ceb2)

**Reply by pw405** [reply]
Oh nice, so it only worked after completely starting over with fresh folder?

**Reply by Reizer88** [reply]
Yes on my pc the SD working after i doing a new clean installation on new folder, but yesterday i have strange experience when i help my brother in law installing SD he using 7800XT and he never use SD before, i got error "No module optimum", then i delete venv folder and run webui-user again and finally after installation the SD is working. I don't know what happen, i still looking for the answer its like random error for me.

**Comment by pw405** [comment]
WTF!! I still can't get this thing to launch!! ARRGGGhhh!! 

`ModuleNotFoundError: No module named 'optimum'`

**Reply by lshqqytiger** [reply]
What did you get when you tried to install optimum manually?
